drivers/counter/counter_native_posix.c

```diff
@@ -32,15 +34,28 @@ static void counter_isr(const void *arg)
 
 	if (is_alarm_pending) {
 		is_alarm_pending = false;
-		pending_alarm.callback(device, 0, current_value,
-				       pending_alarm.user_data);
+		if (pending_alarm.callback) {
+			pending_alarm.callback(device, 0, current_value,
+								   pending_alarm.user_data);
+		}
+	}
+
+	if (is_top_set) {
+		if (current_value == top.ticks) {
+			hw_counter_reset();
+			if (top.callback) {
+				top.callback(device, top.user_data);
+			}
+		}
+		hw_counter_set_target(top.ticks);
 	}
```

Comment:
I think current implementation is OK because it is always cast to `uint32_t` within `counter_native_posix`. And to change this in the way you suggest (setting the ISR to reset the counter when it reaches TOP_VALUE) would require modifying the existing alarm functionality.

Reply:
Nice catch!... Please see latest update in which I've tried to remedy this with the following approaching

* `top` configuration no longer resets the counter value to 0 in the ISR. Only reaching `TOP_VALUE` does this
* After rollover, the next target is the soonest `alarm` or `top` depending on config
* Other cases in which `hw_counter_set_target` is used check whether `TOP_VALUE` will be reached first, and if so, use that as the next target: specifically
  * the `if (is_top_set ...` part of `counter_isr`
  * `ctr_set_alarm`
  * note it is not necessary in `ctr_set_top_value` because either the counter resets, `current_value < ticks`, or error is returned, so no possibility that `TOP_VALUE` is reached before `ticks`